### PR TITLE
set_param

### PR DESCRIPTION
**Pull request by yo96**
More sophisticated `set_param` implementation.

### CODE REVIEW

File: pymtl3/dsl/NamedObject.py [hunk 1]
```diff
@@ -55,13 +56,15 @@ def _construct( s ):
...
-      if "elaborate" in s._dsl.param_dict:
-        more_args = s._dsl.param_dict[ "elaborate" ].iteritems()
+      if "construct" in s._dsl.param_dict:
+        more_args = s._dsl.param_dict[ "construct" ].iteritems()
         kwargs.update( { x: y for x, y in more_args if x } )
 
       s.construct( *s._dsl.args, **kwargs )
 
       s._dsl.constructed = True
+      print( "="*30, "constrcut", "="*30 )
```
Comment (jsn1993) [inline pymtl3/dsl/NamedObject.py]: clean this up

File: pymtl3/passes/test/LineTraceParamPass.py [hunk 2]
```diff
@@ -0,0 +1,48 @@
+"""
+#=========================================================================
+# MultiLevelLineTracePass_test.py
```
Comment (jsn1993) [inline pymtl3/passes/test/LineTraceParamPass.py]: rename

File: pymtl3/passes/LineTraceParamPass.py [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+
+from .BasePass import BasePass, PassMetadata
+
+
+class LineTraceParamPass( BasePass ):
+
+  def __call__( self, top ):
+
+    def wrap_line_trace( obj ):
+      if not hasattr( obj, '_ml_trace' ):
```
Comment (jsn1993) [inline pymtl3/passes/LineTraceParamPass.py]: You might want to assert if you haven't applied this pass already.

File: pymtl3/dsl/NamedObject.py [hunk 4]
```diff
@@ -72,25 +75,41 @@ def __setattr_for_elaborate__( s, name, obj ):
...
+            #     s._dsl.param_dict[ None ].iteritems():
+            #   if compiled_re.match( u_name ):
+            #     for x, y in subdict.iteritems(): # to merge two None subdicts
+            #       if x is None:
+            #         u._dsl.param_dict[ None ].update( subdict )
+            #       else:
+            #         u._dsl.param_dict[ x ] = y
+            
+            # Iterate through the param_dict and update u's dict
+            for key, subdict in s._dsl.param_dict.iteritems():
```
Comment (jsn1993) [inline pymtl3/dsl/NamedObject.py]: Are you going to change this to the new algorithm we discussed yesterday?

Comment (yo96) [reply pymtl3/dsl/NamedObject.py]: Both this and the new algorithm we discussed yesterday still won't work for some corner cases... I just re-implemented the whole thing again.

File: pymtl3/dsl/test/set_param_test.py [hunk 5]
```diff
@@ -0,0 +1,118 @@
...
+  Z = Zoo()
+  Z.set_param( "top.construct", AnimalHouseTypes=[ PandaHouse, TigerTerrace, AnimalHouse, Aquarium ] )
+  Z.set_param( "top.houses[0].construct", AnimalTypes=[ Panda, Panda, Panda   ] )
+  Z.set_param( "top.houses[1].construct", AnimalTypes=[ Tiger, BabyTiger      ] )
+  Z.set_param( "top.houses[2].construct", AnimalTypes=[ Dromaius, HoneyBadger ] )
+  Z.set_param( "top.houses[1].animals[0].construct", MeatTypes=[ Dromaius, HoneyBadger ] )
+  Z.set_param( "top.houses[0].animals*.construct", lunch="bamboo", dinner="bamboo" )
+  Z.set_param( "top.houses[1].animals[0].meats*.construct", lunch="bamboo" )
+  Z.set_param( "top.houses[1].animals[0].meats[1].construct", lunch="grass" )
+  # The following set_param test tricky overwrite behavior.
```
Comment (yo96) [inline pymtl3/dsl/test/set_param_test.py]: This breaks the algorithm we discussed yesterday @jsn1993

**Comment by yo96** [comment]
Cheng and I will add more tests and hopefully some microbenchmarking to evaluate how much the parametrization system slows down elaboration.

File: pymtl3/dsl/test/set_param_test.py (hunk 5 above)
Comment (yo96) [reply pymtl3/dsl/test/set_param_test.py]: line 75~77

Comment (jsn1993) [reply pymtl3/dsl/test/set_param_test.py]: hmm why? the reverse orderof ordededdict still works for each parameter

Comment (jsn1993) [reply pymtl3/dsl/test/set_param_test.py]: hmmm let me think about it. When there are multiple parameters it's not only the name but also the argument name that makes a difference.

**Comment by jsn1993** [comment]
> Cheng and I will add more tests and hopefully some microbenchmarking to evaluate how much the parametrization system slows down elaboration.

sounds good

File: pymtl3/dsl/test/set_param_test.py (hunk 5 above)
Comment (yo96) [reply pymtl3/dsl/test/set_param_test.py]: yes exactly... for line 76, if I delete animals[0] and re-append the new dict, the 'dinner' would be lost
If i pop the node, merge it with the new dict, then re-append it to the end, that still won't work because it screws the order, since that means line 74 may overwrite line 75.

Comment (yo96) [reply pymtl3/dsl/test/set_param_test.py]: In my current algorithm, when inserting a node with '*', it will updates all previous nodes who matches the regex.
It's definitely slower though. I'm trying to do some microbenchmarking to see how much it may slow down the elaboration process.

File: pymtl3/passes/rtlir/rtype/RTLIRType.py [hunk 6]
```diff
@@ -257,10 +257,12 @@ def _gen_parameters( s, obj ):
     arg_names = s.argspec.args[1:]
     assert s.argspec.varargs is None, "varargs are not allowed for construct!"
     assert s.argspec.keywords is None, "keyword args are not allowed for construct!"
-    kwargs = ()
+    # Yanghui : I commented out param_dict related stuff, since we are not
+    # adding extra key word arguments with set_param.
+    # kwargs = ()
```
Comment (yo96) [inline pymtl3/passes/rtlir/rtype/RTLIRType.py]: @ptpan I don't think this part is necessary since `set_param` can only set the value for existing keyword arguments. It won't add extra constructor arguments. Can you take a look at this?

File: pymtl3/passes/rtlir/rtype/test/RTLIRDataType_test.py [hunk 7]
```diff
@@ -31,6 +31,7 @@ def test_py_list():
   with expected_failure( RTLIRConversionError, 'should be a field of some struct' ):
     rdt.get_rtlir_dtype( [ 1, 2, 3 ] )
 
+# Yanghui : this test no longer passes.
 def test_py_struct_arg_no_default_value():
   class B( object ):
     def __init__( s, foo ):
```
Comment (yo96) [inline pymtl3/passes/rtlir/rtype/test/RTLIRDataType_test.py]: @ptpan

File: pymtl3/passes/rtlir/rtype/RTLIRType.py (hunk 6 above)
Comment (ptpan) [reply pymtl3/passes/rtlir/rtype/RTLIRType.py]: Right these are not needed since we can rely on `_dsl.kwargs` to get all keyword argument values.

File: pymtl3/passes/rtlir/rtype/test/RTLIRDataType_test.py (hunk 7 above)
Comment (ptpan) [reply pymtl3/passes/rtlir/rtype/test/RTLIRDataType_test.py]: This branch improves the instantiation of a type and DSL no longer throws the exception this test originally expected. I've updated the test to match the new error message thrown from the RTLIR generation method.

**Comment by yo96** [comment]
@jsn1993 I have finished everything I wanted to do with this PR. Please take a final look.

File: pymtl3/dsl/NamedObject.py [hunk 8]
```diff
@@ -15,19 +15,136 @@
...
 
 class DSLMetadata(object):
   pass
 
+# NOTE: We found that the built-in OrderedDict slows down the elaboration
+# time because much time was spent calling OrderedDict.__init__. 
+# The time for instantiating an OrderedDict is quite long compared
+# to other primitive data structures such as list or dict. We have to
+# implement our own ordered dictionary to mitigate this overhead.
+# TODO: When we move to python3, we won't need this any more.
```
Comment (jsn1993) [inline pymtl3/dsl/NamedObject.py]: great

File: pymtl3/dsl/NamedObject.py [hunk 9]
```diff
@@ -197,16 +307,16 @@ def set_parameter( s, string, value ):
   def _elaborate_construct( s ):
 
     if s._dsl.constructed:
-      print("Don't elaborate the same model twice. \
-             Use APIs to mutate the model.")
+      warnings.warn( "Don't elaborate the same model twice. "
```
Comment (jsn1993) [inline pymtl3/dsl/NamedObject.py]: Interesting. Peitian already mentioned this warning to me. We can discuss whether we want to have warnings throughout this project later. It's fine for now,

File: pymtl3/dsl/test/set_param_test.py [hunk 10]
```diff
@@ -0,0 +1,178 @@
...
+  A.in_ = 1
+  A.tick()
+  print( "A.out==", A.out )
+  assert A.out == 3
+
+  B = IncrArray()
+  B.set_param( "top.construct", DataType=Bits8, num_incrs=3 )
+  B.set_param( "top.incrs*.construct", incr_value=3 )
+  B.set_param( "top.incrs[2].construct", incr_value=5 )
+  B.apply( SimpleSim )
```
Comment (jsn1993) [inline pymtl3/dsl/test/set_param_test.py]: Can you use simple_sim_pass here? We don't want to import stuff from passes/ in dsl.
